processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java

```diff
@@ -58,37 +68,64 @@ public ScanQueryQueryToolChest(
   @Override
   public QueryRunner<ScanResultValue> mergeResults(final QueryRunner<ScanResultValue> runner)
   {
-    return new QueryRunner<ScanResultValue>()
-    {
-      @Override
-      public Sequence<ScanResultValue> run(
-          final QueryPlus<ScanResultValue> queryPlus, final Map<String, Object> responseContext
-      )
-      {
-        // Ensure "legacy" is a non-null value, such that all other nodes this query is forwarded to will treat it
-        // the same way, even if they have different default legacy values.
-        final ScanQuery scanQuery = ((ScanQuery) queryPlus.getQuery()).withNonNullLegacy(scanQueryConfig);
-        final QueryPlus<ScanResultValue> queryPlusWithNonNullLegacy = queryPlus.withQuery(scanQuery);
+    return (queryPlus, responseContext) -> {
+      // Ensure "legacy" is a non-null value, such that all other nodes this query is forwarded to will treat it
+      // the same way, even if they have different default legacy values.
+      final ScanQuery scanQuery = ((ScanQuery) queryPlus.getQuery()).withNonNullLegacy(scanQueryConfig);
+      final QueryPlus<ScanResultValue> queryPlusWithNonNullLegacy = queryPlus.withQuery(scanQuery);
+      final BaseSequence.IteratorMaker scanQueryLimitRowIteratorMaker =
+          new BaseSequence.IteratorMaker<ScanResultValue, ScanQueryLimitRowIterator>()
+          {
+            @Override
+            public ScanQueryLimitRowIterator make()
+            {
+              return new ScanQueryLimitRowIterator(runner, queryPlusWithNonNullLegacy, responseContext);
+            }
 
+            @Override
+            public void cleanup(ScanQueryLimitRowIterator iterFromMake)
+            {
+              CloseQuietly.close(iterFromMake);
+            }
+          };
+
+      if (scanQuery.getTimeOrder().equals(ScanQuery.TIME_ORDER_NONE)) {
         if (scanQuery.getLimit() == Long.MAX_VALUE) {
           return runner.run(queryPlusWithNonNullLegacy, responseContext);
         }
-        return new BaseSequence<>(
-            new BaseSequence.IteratorMaker<ScanResultValue, ScanQueryLimitRowIterator>()
+        return new BaseSequence<ScanResultValue, ScanQueryLimitRowIterator>(scanQueryLimitRowIteratorMaker);
+      } else if ((scanQuery.getTimeOrder().equals(ScanQuery.TIME_ORDER_ASCENDING) ||
+                  scanQuery.getTimeOrder().equals(ScanQuery.TIME_ORDER_DESCENDING))
+                 && scanQuery.getLimit() <= scanQueryConfig.getMaxRowsTimeOrderedInMemory()) {
+        Iterator<ScanResultValue> scanResultIterator = scanQueryLimitRowIteratorMaker.make();
```

Comment:
This looks like it will apply the limit before sorting takes place. Am I reading it right? It needs to be the other way around, otherwise we won't be guaranteed to get the very earliest or very latest rows.

Also, for a Scan query with a limit, in order to make sure we read the very earliest or very latest rows, we need to make sure to iterate through segments in the right order. Otherwise we are going to need to read the first (or last) `limit` rows out of _every_ segment to be sure we got the right ones, which we want to avoid. The idea should be to read segments in ascending (or descending) order, and stop reading once we know that no as-yet-unread segments could possibly offer us any earlier (or later) events, based on those segments' data intervals.

Reply:
Ah shoot, I misunderstood part of the issue.  I wrote this under the assumption that it would act like an ORDER BY operator.  So if I understand correctly now, this query:

```
{
  ...
  "timeOrdering":"descending" 
  "limit":100
}
```

should return the latest 100 rows?

Conversely, this query:

```
{
  ...
  "timeOrdering":"ascending" 
  "limit":100
}
```

should return the earliest 100 rows?

Reply:
Yes, that's correct. It is how `ORDER BY` works in SQL, too (ordering happens before limiting).